src/storage/src/controller.rs

```diff
@@ -703,38 +825,104 @@ where
+            .get_mut(&id)
+            .ok_or(StorageError::IdentifierMissing(id))
+    }
+
     async fn create_exports(
         &mut self,
-        exports: Vec<(GlobalId, ExportDescription)>,
+        exports: Vec<(GlobalId, ExportDescription<Self::Timestamp>)>,
     ) -> Result<(), StorageError> {
+        // Validate first, to avoid corrupting state.
+        let mut dedup_hashmap = HashMap::<&_, &_>::new();
+        for (id, desc) in exports.iter() {
+            if dedup_hashmap.insert(id, desc).is_some() {
+                return Err(StorageError::SourceIdReused(*id));
+            }
+            if let Ok(export) = self.export(*id) {
+                if &export.description != desc {
+                    return Err(StorageError::SourceIdReused(*id));
+                }
+            }
+        }
+
         for (id, description) in exports {
-            let augmented_sink_desc = SinkDesc {
-                from: description.sink.from,
-                from_desc: description.sink.from_desc,
-                connection: match description.sink.connection {
-                    SinkConnection::Kafka(k) => SinkConnection::Kafka(k),
-                    SinkConnection::Tail(t) => SinkConnection::Tail(t),
-                    SinkConnection::Persist(PersistSinkConnection {
-                        value_desc,
-                        storage_metadata: (),
-                    }) => SinkConnection::Persist(PersistSinkConnection {
-                        value_desc,
-                        storage_metadata: self.collection(id)?.collection_metadata.clone(),
-                    }),
-                },
-                envelope: description.sink.envelope,
-                // TODO(chae): derive this (sink_description.as_of is hardcoded to u64 instead of general timestamp T)
-                as_of: SinkAsOf {
-                    frontier: Antichain::from_elem(T::minimum()),
-                    strict: description.sink.as_of.strict,
-                },
+            let from = description.sink.from;
+
+            self.state
+                .exports
+                .insert(id, ExportState::new(description.clone()));
+
+            self.state
+                .exported_collections
+                .entry(from)
+                .or_default()
+                .push(id);
+
+            let from_storage_metadata = self.collection(from)?.collection_metadata.clone();
+
+            let initial_as_of = MetadataExportFetcher::get_stash_collection()
+                .insert_without_overwrite(
+                    &mut self.state.stash,
+                    &id,
+                    DurableExportMetadata {
+                        initial_as_of: description.sink.as_of,
+                    },
+                )
+                .await?
+                .initial_as_of;
+
+            // We've added the dependency above in `exported_collections` so this guaranteed not to change at least until the sink is started up.
+            let from_since = self
```

Comment:
Have you tried using sth like this?:
```
            let from_since = from_collection.implied_capability;
```

The `implied_capability` is initialized from `ReadHandle::since` when creating the collection states, so it should be the same. And if the controller is doing it's job correctly, the `since` of the shard will not have advanced beyond that.

Reply:
Let me look into this! That would be simpler.  I think I, without looking into it too much, decided to originally skip this because I was worried it was possible for that value to be a little stale -- but you're right that this would simplify a lot!